daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Utf8String.scala

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.daml.lf.data
+
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+
+import scalaz.Equal
+import scalaz.std.string._
+
+import scala.collection.mutable
+
+// The DAML-LF strings are supposed to be UTF-8 while standard java string are UTF16
+// We box standard UTF16 java String to prevent non intentional usages of UTF16 operations
+// (for instance length, charAt, ordering ...) and provide UTF8 emulation methods.
+
+// Use `toString` wisely. As a rule on the thumb, you should use `toString` only for logging,
+// testing, or sending to external libraries (as for instance protobuf/json builders).
+case class Utf8String(override val toString: String) extends AnyVal with Ordered[Utf8String] {
+
+  // The DAML-LF strings are supposed to be UTF-8.
+  // However standard "exploding" java/scala methods like
+  // _.toList split in Character which are not Unicode codepoint.
+  def explode: ImmArray[Utf8String] = {
+    val len = toString.length
+    val arr = new mutable.ArraySeq[Utf8String](toString.codePointCount(0, len))
```

Comment:
```suggestion
    val arr = ImmArray.newBuilder[Utf8String]
    arr.sizeHint(javaString.codePointCount(0, len))
```

Then `arr += Utf8...` and `arr.result()` below to fix up the types.

Reply:
fixed